File: DevelopmentUtility/youyou_Proto/youyou_Proto/bin/Debug/CreateProto/TestProto.cs
//===================================================
//作    者：
//创建时间：2022-03-18 14:54:31
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 测试协议
/// </summary>
public struct TestProto : IProto
{
    public ushort ProtoCode { get { return 1004; } }

    public int Id; //编号
    public string Name; //名称
    public int Type; //类型
    public float Price; //价格

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(Id);
            ms.WriteUTF8String(Name);
            ms.WriteInt(Type);
            ms.WriteFloat(Price);
            return ms.ToArray();
        }
    }

    public static TestProto GetProto(byte[] buffer)
    {
        TestProto proto = new TestProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.Id = ms.ReadInt();
            proto.Name = ms.ReadUTF8String();
            proto.Type = ms.ReadInt();
            proto.Price = ms.ReadFloat();
        }
        return proto;
    }
}

File: DevelopmentUtility/youyou_Proto/youyou_Proto/bin/Debug/CreateLuaProto/TestProto.lua
--测试协议
TestProto = { ProtoCode = 1004, Id = 0, Name = "", Type = 0, Price = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
TestProto.__index = TestProto;

function TestProto.New()
    local self = { }; --初始化self
    setmetatable(self, TestProto); --将self的元表设定为Class
    return self;
end


--发送协议
function TestProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.Id);
    ms:WriteUTF8String(proto.Name);
    ms:WriteInt(proto.Type);
    ms:WriteFloat(proto.Price);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function TestProto.GetProto(buffer)

    local proto = TestProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.Id = ms:ReadInt();
    proto.Name = ms:ReadUTF8String();
    proto.Type = ms:ReadInt();
    proto.Price = ms:ReadFloat();

    ms:Dispose();
    return proto;
end